File: leetcode_js/PlusOne.js
/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
    for (let i = digits.length - 1; i >= 0; i--) {
        // Normal case (no carry over)
        // Just increment the last element in the array by 1
        if (digits[i] < 9) {
            digits[i] = digits[i] + 1;
            return digits;
        } else {
            digits[i] = 0;

        }
    }
    
    // Edge case where all elements are 9s
    digits.unshift(1)
    return digits;
};

// Tests
console.log(plusOne([1,2,3]));
console.log(plusOne([0, 0, 0]));
console.log(plusOne([1, 2, 9]));

File: leetcode_py/PlusOne.py
from typing import List

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        for i in range(len(digits) - 1, -1, -1):
            print(digits[i])
            # Increment number and return
            if digits[i] < 9:
                digits[i] = digits[i] + 1
                return digits
            else:
                digits[i] = 0
        # all 9 case
        digits.insert(0, 1)
        return digits

# Tests
example = Solution()
# example.plusOne([9,9,9])
print(example.plusOne([1,2,3]))
# example.plusOne([1,2,9])
